### PR TITLE
[None][fix] Clamp small non-zero sampling temperature

### LINKED ISSUE #15715: [Bug]: Clamp very small non-zero temperature values to avoid numerical instability
### System Info

CPU architecture: x86_64

CPU: Intel Xeon Gold 6462C, 8 vCPUs (4 cores × 2 threads), max 3900 MHz

GPU name: NVIDIA L20
GPU memory size: 48G

TensorRT-LLM branch or tag: main
TensorRT-LLM commit: https://github.com/NVIDIA/TensorRT-LLM/commit/28acbad14b442e89312cfe9e845a5f721a3e702f
PyTorch: 2.10.0+cu130
CUDA: 13.0.88
cuDNN: 9.15.0
NVIDIA driver version: 550.163.01

OS: Ubuntu 24.04.3 LTS

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

```python
from tensorrt_llm.sampling_params import SamplingParams

# Before the fix: very small non-zero temperature bypasses validation
# and enters the sampling backend, causing potential inf/nan in logits/temperature
sp = SamplingParams(temperature=1e-12, top_k=50)
print(sp.temperature)  # prints 1e-12 — dangerously small

# Simulating what happens inside the backend:
import torch
logits = torch.tensor([1.0, 2.0, 3.0])
temperature = 1e-12
scaled = logits / temperature
print(scaled)  # tensor([1.0e+12, 2.0e+12, 3.0e+12]) → leads to inf/nan in softmax
print(torch.softmax(scaled, dim=-1))  # tensor([0., 0., 1.]) or nan depending on dtype
```

### Expected behavior

SamplingParams should reject or safely normalize temperature values that are 
positive but below a numerical stability threshold.

Specifically:
- temperature=0.0  → kept as 0.0 (greedy decoding)
- temperature=1e-12 → clamped to MIN_SAMPLING_TEMPERATURE (1e-2)
- temperature=1.0  → kept as 1.0 (normal sampling)

### actual behavior

amplingParams only validates temperature >= 0, so values like 1e-12 
pass validation unchanged and are forwarded to the sampling backend.

Inside the backend, logits / temperature with temperature=1e-12 produces 
extremely large logit values (e.g., 1e+12), which overflow to inf or nan 
in float16/bfloat16 precision, leading to undefined sampling behavior.

Example:
```python
import torch
logits = torch.tensor([1.0, 2.0, 3.0], dtype=torch.float16)
scaled = logits / 1e-12
print(scaled)
# tensor([inf, inf, inf], dtype=torch.float16)

print(torch.softmax(scaled, dim=-1))
# tensor([nan, nan, nan], dtype=torch.float16)
```

### additional notes

1.  This fix mirrors the same numerical stability guard used by vLLM    (temperature = max(temperature, 1e-2) for non-zero values).

2. The issue is most severe in float16/bfloat16: typical logit values (~10)    divided by 1e-6 already exceed float16's max range (~65504),    producing inf/nan in softmax.

3. temperature=0.0 (greedy decoding) is intentionally excluded from the clamp    and its behavior is unchanged.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and checked the [documentation](https://nvidia.github.io/TensorRT-LLM/) and [examples](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples) for answers to frequently asked questions.

### PR DESCRIPTION
**Pull request by chfeng-cs**
## Summary by CodeRabbit

* **Bug Fixes**
  * Improved temperature validation to prevent very small nonzero values from causing unstable sampling behavior.
  * Very small temperatures are now automatically adjusted up to a safe minimum value.
  * Added coverage to verify the adjusted temperature is reflected correctly in sampling output.

## Description

Fix #15715 .This PR clamps very small non-zero `SamplingParams.temperature` values to `1e-2` before they are passed to the sampling backend.

`temperature=0` continues to represent greedy decoding, while positive values below the stability threshold are normalized to avoid excessive logit scaling from `logits / temperature`.

TensorRT-LLM currently validates that `temperature >= 0`, but values such as `1e-12` can still enter the sampling path and may lead to `inf`, `nan`, or unstable sampling behavior in edge cases.

This mirrors the numerical stability guard used by vLLM while preserving the existing greedy decoding semantics for `temperature=0`.

# Changes

Added `MIN_SAMPLING_TEMPERATURE = 1e-2`.

Clamped `0 < temperature < MIN_SAMPLING_TEMPERATURE` to `MIN_SAMPLING_TEMPERATURE` in `SamplingParams._validate()`.

Added debug logging when a small non-zero temperature is clamped.

Added a unit test covering `temperature=0`, very small positive temperatures, the threshold value, and normal temperatures.

## Test Coverage

```shell
pytest tests/unittest/llmapi/test_sampling_params.py -k temperature_clamps_small_nonzero_values
```

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [ ] Please check this after reviewing the above items as appropriate for this PR.

### CODE REVIEW

File: tensorrt_llm/sampling_params.py [hunk 1]
```diff
@@ -372,6 +373,12 @@ def _validate(self):
             raise ValueError(f"require top_k >= 0, got top_k={self.top_k}")
         if self.temperature is not None and self.temperature < 0:
             raise ValueError(f"require temperature >= 0, got temperature={self.temperature}")
+        if self.temperature is not None and 0 < self.temperature < MIN_SAMPLING_TEMPERATURE:
+            logger.debug(
+                f"Clamping temperature from {self.temperature} to {MIN_SAMPLING_TEMPERATURE} "
+                "to avoid numerical instability."
+            )
+            self.temperature = MIN_SAMPLING_TEMPERATURE
```
[the remark this answers is not in the document]
Comment (chfeng-cs) [reply tensorrt_llm/sampling_params.py]: done

**Comment by chfeng-cs** [comment]
@YihuiLu512 @lori-ren, could you please review #15716 before the merge decision and consider consolidating the work here, since this was the earlier implementation for #15715?

**Comment by zhaoyangwang-nvidia** [comment]
Thanks for the fix! Before we settle on the approach, a few questions:

The sampler already handles this case internally: GREEDY_TEMPERATURE_THRESHOLD = 1e-4 in _torch/pyexecutor/sampler/ops/vanilla.py treats any request with T <= 1e-4 as greedy (the temperature is never used as a divisor), so on the PyTorch sampler path a value like 1e-12 safely degrades to greedy rather than overflowing.

Given that, I have some concerns about clamping up to MIN_SAMPLING_TEMPERATURE = 1e-2:

- It works against user intent. Someone passing T = 1e-12 clearly wants near-deterministic output. Clamping to 1e-2 silently turns that into genuine sampling, whereas snapping small values down to greedy (like the existing sampler-side threshold, and IIRC also what vLLM does with its epsilon check) preserves the intent and avoids the division entirely.
- It silently changes valid requests. T = 5e-3 is numerically fine to compute, but would be rewritten to 1e-2 — a visibly different distribution — with only a debug-level log.
- Two disconnected thresholds. After this change we'd have 1e-2 at the API layer and 1e-4 inside the sampler, unaware of each other; the sampler-side one becomes dead code for API users.

Could you share which path actually exhibited the inf/nan in #15715 (TRTLLMSampler? a path that bypasses the greedy threshold?) That would help decide whether the right fix is mapping tiny temperatures to greedy at the API level (reusing/aligning with GREEDY_TEMPERATURE_THRESHOLD), or adding the same guard to the affected path, rather than introducing a new clamp value.

**Comment by chfeng-cs** [comment]
@zhaoyangwang-nvidia Thanks for raising these questions.  I appreciate the concerns you raised. I’ll carefully review the current sampling paths, including the exact coverage of the existing threshold and which paths may still be affected, and then follow up with a detailed answer.

**Comment by zhaoyangwang-nvidia** [comment]
> @zhaoyangwang-nvidia Thanks for raising these questions. I appreciate the concerns you raised. I’ll carefully review the current sampling paths, including the exact coverage of the existing threshold and which paths may still be affected, and then follow up with a detailed answer.

Thanks for the context and for double-checking the current sampling paths.

Just to add, we are also actively refactoring the Torch sampler recently, so some of the related logic may continue to evolve. If you find any bugs, corner cases, or have suggestions while reviewing the sampling behavior, please feel free to raise them or discuss with me directly. I’d be happy to go through them together.

**Comment by mikeiovine** [comment]
Closing as no activity for 4 weeks, please feel free to reopen if you'd like to have it reviewed!
